### PR TITLE
chore(theme): SASS-based themes

### PR DESCRIPTION
**Pull request by redox**
What do you think about SASS-based themes?

TODO:
 * [x] color variables
 * [x] autoprefixer -> we'll use bourbon/compass if required
 * [x] npm shrinkwrap --dev

Ref #290

### CODE REVIEW

**Comment by vvo** [comment]
Never used sass and would prefer to use postcss as proposed in #290.

We can do the exact same PR and see how it goes with postcss.

**Comment by redox** [comment]
> Never used sass and would prefer to use postcss as proposed in #290.

Happy to compare once someone give postcss a try ;)

**Comment by redox** [comment]
> We can do the exact same PR and see how it goes with postcss.

Before going for SASS I tried PostCSS and couldn't make it work the way I was imagining it -> so I fallback to what I knew :)

**Comment by vvo** [comment]
> Happy to compare once someone give postcss a try ;)

This will be a :vs: :video_game:

**Comment by Jerskouille** [comment]
Should I do a `stylus` PR ? :D

File: themes/_base.sass [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  .ais-#{$component}
+    @content
+
+@mixin modifier($modifier)
+  &--#{$modifier}
+    @content
+
+@mixin element($element)
+  &__#{$element}
+    @content
```
Comment (pixelastic) [inline themes/_base.sass]: How would you use that?

Hmm, the goal of BEM is to not have to do any nesting. Sass can let you easily do a lot of nesting, but we should actually never nest more than 2 or 3 levels deep.

**Comment by pixelastic** [comment]
I wanted to do some Sass at first but @vvo had a very good point about writing CSS for the future. We would have used a postprocessor for cross-compat anyway, so I'm gonna see if all the variables/mixins can still be replicated in postcss

**Comment by vvo** [comment]
Careful, we said in #290 that we were investigating this but that should not even be needed for a 1.0.0 release. Themes can just be jsdelivr/default.css+theme.css and that's fine like that.

File: themes/_base.sass (hunk 1 above)
Comment (redox) [inline themes/_base.sass]: Hmmm, the good thing it that is doesn't nest anything :) The nesting is just use to build the class names from the parent using that `&--` and `&__` syntax.

Comment (pixelastic) [inline themes/_base.sass]: Great. I did not know that syntax.

I'm not sure the end result in Sass will be any more readable than plain CSS, though :)

**Comment by redox** [comment]
> Careful, we said in #290 that we were investigating this but that should not even be needed for a 1.0.0 release. Themes can just be jsdelivr/default.css+theme.css and that's fine like that.

If we want the community to build theme, we'll need to. For 1.0 I'm happy using SASS and move to anything else if you think it's better.

Regarding the cross-compat, that has been solved years ago with autoprefixer & cie; I'm really not afraid of that. To be honest, I don't see the benefit of PostCSS compared to SASS for what we need.

File: themes/_base.sass (hunk 1 above)
Comment (redox) [inline themes/_base.sass]: > I'm not sure the end result in Sass will be any more readable than plain CSS, though :)

For me, no-brainer :)

```css
/* REFINEMENT LIST */
.ais-refinement-list {
}
.ais-refinement-list--header {
  font-size: 1.2em;
}
.ais-refinement-list--body {
}
.ais-refinement-list--list {
}
.ais-refinement-list--item {
}
.ais-refinement-list--item__active {
  font-weight: bold;
}
.ais-refinement-list--label {
}
.ais-refinement-list--checkbox {
}
.ais-refinement-list--count {
  float: right;
  display: inline-block;
  background: black;
  color: white;
}
.ais-refinement-list--footer {
}
```

VS 

```sass
// REFINEMENT LIST
+component(refinement-list)
  +modifier(header)
    font-size: 1.2em
  +modifier(body)
  +modifier(list)
  +modifier(item)
    +element(active)
       font-weight: bold
  +modifier(label)
  +modifier(checkbox)
  +modifier(count)
    float: right
    display: inline-block
    background: black
    color: white
  +modifier(footer)
```

**Comment by pixelastic** [comment]
> To be honest, I don't see the benefit of PostCSS compared to SASS for what we need.

I would say, the same as using babel instead of coffeescript. Write for the future, in something close to a standard. Feature-wise it will not change much, so ok to do a Sass version first if it's what we're more proficient in.

**Comment by vvo** [comment]
> Sass version first if it's what we're more proficient in.

I cannot read any SASS but that's ok if I do not need to :)

**Comment by vvo** [comment]
> If we want the community to build theme, we'll need to.

Right now if you need to build a theme you have to include default.css and then only override the classes you need. Even using SASS or postcss feels to early for me, we haven't built any theme over default.css

**Comment by redox** [comment]
> Even using SASS or postcss feels too early for me, we haven't built any theme over default.css

Actually, from my POV that's something required to build a theme. We're so used to pre-processors than going back to vanilla CSS is like going back to ES3 :)

**Comment by vvo** [comment]
> Actually, from my POV that's something required to build a theme. We're so used to pre-processors than going back to vanilla CSS is like going back to ES3 :)

Yes I get it, I never used any pre processor to do any CSS file. Only small css files and utilities.

**Comment by bobylito** [comment]
Not convinced by SASS overall and would prefer a more future proof / flexible approach. I do admit that it seems powerful though. So up to you.

@redox As for the comparaison with ES3, I would say using SASS now feels like starting using coffeescript when ES6 is out...

**Comment by redox** [comment]
> @redox As for the comparaison with ES3, I would say using SASS now feels like starting using coffeescript when ES6 is out...

Not really, SASS is a really a common use-case: a big majority of front projects are using S{C,A}SS (including bootstrap v4 for instance); on the other hand a very minority of JS projects are using Coffee.

**Comment by Shipow** [comment]
I must admit postCSS looks nice, but it also looks a bit more complex and less readable. I haven't tried it yet, who did?
So I'm happy with SASS (surprise) for the first versions.

**Comment by vvo** [comment]
> I haven't tried it yet, who did?

I would like too, we should try to find 1-2 hours to at least try to reproduce what @redox did with postcss + plugins (autoprefixer, mixins).

Also I would like to see what a theme would look like in term of reutilization, I guess @pixelastic you are on it?

File: themes/_base.sass (hunk 1 above)
Comment (pixelastic) [inline themes/_base.sass]: Yeah, my point exactly. I find the first one easier to read :)

Comment (redox) [inline themes/_base.sass]: So you can use the other syntax:

```sass
// REFINEMENT LIST
+bem(refinement-list, header)
  font-size: 1.2em
+bem(refinement-list, body)
+bem(refinement-list, list)
+bem(refinement-list, header)
+bem(refinement-list, item)
+bem(refinement-list, item, active)
     font-weight: bold
+bem(refinement-list, label)
+bem(refinement-list, checkbox)
+bem(refinement-list, count)
    float: right
    display: inline-block
    background: black
    color: white
+bem(refinement-list, footer)
```

**Pull request merged**